feat(cudf): Implement and default to automatic GPU detection

zoltan opened the pull request:
Instead of providing an enabled configuration option, we're explicitly enabling cuDF with "auto" mode.

New behavior:
- unset (default): auto-detect GPU availability, enable cuDF if GPU is present
- cudf.disabled=false: require cuDF, fail if no GPU is present
- cudf.disabled=true: force disable cuDF

This is a QoL improvement for mixed cluster deployments and enables
shipping a single image that works on systems with and without a GPU.

majetideepak commented on velox/experimental/cudf/exec/ToCudf.cpp:
```diff
@@ -491,7 +503,20 @@ CudfConfig& CudfConfig::getInstance() {
 void CudfConfig::initialize(
     std::unordered_map<std::string, std::string>&& config) {
   if (config.find(kCudfEnabled) != config.end()) {
-    enabled = folly::to<bool>(config[kCudfEnabled]);
```
We have to change the config type to string.
I think the default semantics should be like auto unless explicitly disabled. 
So convert `enabled` -> `disabled`. disabled set to false or unset means auto.

zoltan replied:
but of course to change _the default_, it needs to become a string.

zoltan committed:
feat(cudf): Implement and default to automatic GPU detection

Instead of providing an enabled configuration option, we're explicitly
enabling cuDF with "auto" mode.

New behavior:
- unset (default): auto-detect GPU availability, enable cuDF if GPU is present
- cudf.disabled=false: require cuDF, fail if no GPU is present
- cudf.disabled=true: force disable cuDF

This is a QoL improvement for mixed cluster deployments and enables
shipping a single image that works on systems with and without a GPU.

lga-zurich commented on velox/experimental/cudf/exec/ToCudf.cpp:
```diff
@@ -420,8 +432,8 @@ struct CudfDriverAdapter {
-            CudfConfig::kCudfEnabled, CudfConfig::getInstance().enabled) &&
+    if (driver.driverCtx()->queryConfig().get<bool>(
+            CudfConfig::kCudfDisabled, CudfConfig::getInstance().disabled) &&
         allowCpuFallback_) {
```
If cudf is disabled, do we care whether cpu fallback is allowed or not? That will just make for another combination of conflicting configurations. I would say that if cudf is not enabled, the rest of the cudf parameters should be irrelevant.

karthikeyann commented:
@simoneves after this PR is merged, config templates in https://github.com/rapidsai/velox-testing/blob/main/presto/docker/config/template/etc_worker/config_native.properties should be updated

jinchengchenghh commented on velox/docs/configs.rst:
```diff
@@ -1133,10 +1133,10 @@ Note: These configurations are experimental and subject to change.
-     - bool
-     - true
-     - If true, enable cuDF. By default, it is enabled if compiled with cuDF.
+   * - cudf.disabled
```
Why not still use cudf.enabled with default true value? And always detect if gpu is available, if not available, fallback to cpu, and log a warning

lga-zurich replied:
> Double negatives do hurt my brain

I also find it very confusing.

Also agree that having a `.disabled` parameter makes it feel like something you need to opt-out of.

Finally, if the user does set it to some value, I don't think `auto` should second-guess the user and make it run where it shouldn't.

zoltan replied:
ok, how about I switch back the enabled, and then
* if it's unset, we'll default to auto
* if it's enabled, and we don't have a GPU, we'll fail
* if it's disabled, it's disabled
would you be all OK with that?

@devavret as for opt-in: the user already opts-in because the VELOX_ENABLE_CUDF is off by default, in my opinion. if you're compiling with cuDF *and not setting* enabled to anything explicitly, I feel like auto is a good choice. wdyt?

zoltan replied:
the current code does not detect it for each task/plan. it is detected at startup, once.

jinchengchenghh commented on velox/experimental/cudf/exec/ToCudf.cpp:
```diff
@@ -64,6 +64,18 @@ bool isAnyOf(const Base* p) {
 
+bool detectGpuAvailable() {
+  int deviceCount = 0;
+  auto error = cudaGetDeviceCount(&deviceCount);
```
If this runs in a CPU node, does cuda install?
